sessionctx/stmtctx/stmtctx.go

```diff
@@ -611,13 +612,34 @@ func (sc *StatementContext) SetPlanHint(hint string) {
 	sc.planHint = hint
 }
 
+// PlanCacheType is the flag of plan cache
+type PlanCacheType int
+
+const (
+	// DefaultNoCache no cache
+	DefaultNoCache PlanCacheType = iota
+	// SessionPrepared session prepared plan cache
+	SessionPrepared
+	// SessionNonPrepared session non-prepared plan cache
+	SessionNonPrepared
+)
+
 // SetSkipPlanCache sets to skip the plan cache and records the reason.
-func (sc *StatementContext) SetSkipPlanCache(reason error) {
+func (sc *StatementContext) SetSkipPlanCache(cacheType PlanCacheType, reason error) {
 	if !sc.UseCache {
 		return // avoid unnecessary warnings
 	}
 	sc.UseCache = false
-	sc.AppendWarning(reason)
+	switch cacheType {
+	case DefaultNoCache:
+		sc.AppendWarning(errors.New("unknown cache type"))
+	case SessionPrepared:
+		sc.AppendWarning(errors.Errorf("skip prepared plan-cache: %s", reason.Error()))
+	case SessionNonPrepared:
+		if sc.InExplainStmt {
```

Comment:
Do we need to check the explain format here? Only when it is in 'format=plan_cache', we need to output a warning here.

Reply:
yes, i think we need this, some type convert check will be doing not at the explain statement, if we don't check this, we may get warnings when execute a query such as:
```sql
mysql> select * from t where a=1.1;
Empty set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------------------------------------+
| Level   | Code | Message                                                     |
+---------+------+-------------------------------------------------------------+
| Warning | 1105 | skip non-prepared plan-cache: '1.1' may be converted to INT |
+---------+------+-------------------------------------------------------------+
1 row in set (0.01 sec)
```

Reply:
if we don't check this, we may need to add some judgment logic at the `SetSkipPlanCache`call place